File: library/src/androidTest/java/com/doctoror/particlesdrawable/ParticlesDrawableTest.java
/*
 * Copyright (C) 2017 Yaroslav Mytkalyk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.doctoror.particlesdrawable;

import org.junit.Test;

import android.annotation.TargetApi;
import android.content.res.Resources;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.test.InstrumentationRegistry;
import android.util.TypedValue;

import static org.junit.Assert.*;

/**
 * {@link ParticlesDrawable} test
 */
public final class ParticlesDrawableTest {

    @Test
    public void testIsRunningByDefault() {
        assertFalse(new ParticlesDrawable().isRunning());
    }

    @Test
    public void testIsRunningWhenStarted() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.setBounds(0, 0, 10, 10);
        d.start();
        try {
            assertTrue(d.isRunning());
        } finally {
            d.stop();
        }
    }

    @Test
    public void testIsStopedWhenStopped() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.setBounds(0, 0, 10, 10);
        d.start();
        d.stop();
        assertFalse(d.isRunning());
    }

    @Test
    public void testSetBoundsWhenRunning() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.start();
        d.setBounds(0, 0, 10, 10);
        d.stop();
    }

    @Test
    public void testWithZeroBounds() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.setBounds(0, 0, 0, 0);
        d.start();
        assertTrue(d.isRunning());
        d.stop();
        assertFalse(d.isRunning());
    }

    @Test
    public void testMakeBrandNewFrameWithZeroBounds() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.makeBrandNewFrame();
    }

    @Test
    public void testMakeBrandNewFrame() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.setBounds(0, 0, 10, 10);
        d.makeBrandNewFrame();
    }

    @Test
    public void testMakeBrandNewFrameWhenRunning() {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.setBounds(0, 0, 10, 10);
        d.start();
        d.makeBrandNewFrame();
        d.stop();
    }

    @Test
    public void testDotRadiusRangeThatRoundsUpToTheSameInt() throws Throwable {
        final ParticlesDrawable d = new ParticlesDrawable();
        d.setBounds(0, 0, 10, 10);
        d.setDotRadiusRange(0.5f, 0.6f);
        d.makeBrandNewFrame();
    }

    @NonNull
    @RequiresApi(Build.VERSION_CODES.N)
    private ParticlesDrawable inflateCustomizedDrawable() {
        final ParticlesDrawable result = (ParticlesDrawable) InstrumentationRegistry.getContext()
                .getDrawable(com.doctoror.particlesdrawable.test.R.drawable.drawable_customized);
        if (result == null) {
            throw new NullPointerException("getDrawable() returned null");
        }
        return result;
    }

    private float dp(final float value) {
        return TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP, value, Resources.getSystem().getDisplayMetrics());
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testDotRadiusRangeFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(dp(2), v.getMinDotRadius(), Config.ASSERT_DELTA);
        assertEquals(dp(3), v.getMaxDotRadius(), Config.ASSERT_DELTA);
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testLineThicknessFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(dp(4), v.getLineThickness(), Config.ASSERT_DELTA);
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testLineDistanceFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(dp(5), v.getLineDistance(), Config.ASSERT_DELTA);
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testNumDotsFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(6, v.getNumDots());
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testDotColorFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(0xfffffff0, v.getDotColor());
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testLineColorFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(0xfffffff1, v.getLineColor());
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testFrameDelayFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(7, v.getFrameDelay());
    }

    @Test
    @TargetApi(Build.VERSION_CODES.N)
    public void testStepMultiplierFromAttributes() {
        final ParticlesDrawable v = inflateCustomizedDrawable();
        assertEquals(1.1, v.getStepMultiplier(), Config.ASSERT_DELTA);
    }

    @Test
    public void testSetFrameDelay() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setFrameDelay(1);
        assertEquals(1, s.getFrameDelay());
    }

    @Test
    public void testSetStepMultiplier() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setStepMultiplier(0f);
        assertEquals(0, s.getStepMultiplier(), Config.ASSERT_DELTA);
    }

    @Test
    public void testSetDotRadiusRange() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setDotRadiusRange(0.5f, 0.6f);
        assertEquals(0.5, s.getMinDotRadius(), Config.ASSERT_DELTA);
        assertEquals(0.6, s.getMaxDotRadius(), Config.ASSERT_DELTA);
    }

    @Test
    public void testSetLineThickness() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setLineThickness(1f);
        assertEquals(1, s.getLineThickness(), Config.ASSERT_DELTA);
    }

    @Test
    public void testSetLineDistance() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setLineDistance(0f);
        assertEquals(0, s.getLineDistance(), Config.ASSERT_DELTA);
    }

    @Test
    public void testSetNumDots() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setNumDots(0);
        assertEquals(0, s.getNumDots());
    }

    @Test
    public void testSetLineColor() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setLineColor(2);
        assertEquals(2, s.getLineColor());
    }

    @Test
    public void testSetDotColor() {
        final ParticlesDrawable s = new ParticlesDrawable();
        s.setDotColor(3);
        assertEquals(3, s.getDotColor());
    }
}

File: library/src/test/java/com/doctoror/particlesdrawable/ParticlesDrawableTest.kt
/*
 * Copyright (C) 2018 Yaroslav Mytkalyk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.doctoror.particlesdrawable

import android.content.res.Resources
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.PixelFormat
import android.os.Build
import android.os.SystemClock
import android.util.AttributeSet
import com.doctoror.particlesdrawable.engine.Engine
import com.doctoror.particlesdrawable.engine.SceneConfigurator
import com.doctoror.particlesdrawable.model.Scene
import com.doctoror.particlesdrawable.renderer.CanvasSceneRenderer
import com.nhaarman.mockito_kotlin.*
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.xmlpull.v1.XmlPullParser
import java.lang.reflect.Modifier

@Config(manifest = Config.NONE, sdk = [Build.VERSION_CODES.LOLLIPOP])
@RunWith(RobolectricTestRunner::class)
class ParticlesDrawableTest {

    private val canvasRenderer: CanvasSceneRenderer = mock()
    private val scene: Scene = mock()
    private val sceneConfigurator: SceneConfigurator = mock()
    private val engine: Engine = mock()

    private val underTest = spy(ParticlesDrawable().apply {
        overrideMember(this, "canvasRenderer", canvasRenderer)
        overrideMember(this, "scene", scene)
        overrideMember(this, "sceneConfigurator", sceneConfigurator)
        overrideMember(this, "engine", engine)
    })

    private fun overrideMember(target: Any, name: String, value: Any) {
        val field = target::class.java.getDeclaredField(name)
        field.isAccessible = true

        val modifiers = field.javaClass.getDeclaredField("modifiers")
        modifiers.isAccessible = true
        modifiers.setInt(field, field.modifiers and Modifier.FINAL.inv())

        field.set(target, value)
    }

    @Test
    fun inflates() {
        val attrs: AttributeSet = mock()
        val typedArray: TypedArray = mock()
        val resources: Resources = mock {
            on(it.obtainAttributes(eq(attrs), any())).thenReturn(typedArray)
        }

        val xmlPullParser: XmlPullParser = mock()

        underTest.inflate(resources, xmlPullParser, attrs, null)

        verify(sceneConfigurator).configureSceneFromAttributes(scene, resources, attrs)
    }

    @Test
    fun setsDimensionsOnBoundsChange() {
        val width = 240
        val height = 320

        underTest.setBounds(0, 0, width, height)

        verify(engine).setDimensions(width, height)
    }

    @Test
    fun drawsAndRuns() {
        val canvas: Canvas = mock()

        underTest.draw(canvas)

        val inorder = inOrder(canvasRenderer, engine)
        inorder.verify(canvasRenderer).setCanvas(canvas)
        inorder.verify(engine).draw()
        inorder.verify(canvasRenderer).setCanvas(null)
        inorder.verify(engine).run()
    }

    @Test
    fun invalidatesSelfOnRequestRender() {
        underTest.requestRender()
        verify(underTest).invalidateSelf()
    }

    @Test
    fun reqeustsRenderOnScheduleNextFrameWithZeroDelay() {
        underTest.scheduleNextFrame(0)
        verify(underTest).requestRender()
    }

    @Test
    fun schedulesSelfOnScheduleNextFrameWithNonZeroDelay() {
        val delay = 36000L

        underTest.scheduleNextFrame(delay)

        val timeCaptor = argumentCaptor<Long>()
        verify(underTest).scheduleSelf(any(), timeCaptor.capture())

        assertEquals(SystemClock.uptimeMillis() + delay, timeCaptor.firstValue)
    }

    @Test
    fun unschedulesSelfOnUnscheduleNextFrame() {
        underTest.unscheduleNextFrame()
        verify(underTest).unscheduleSelf(any())
    }

    @Test
    fun setsAlphaToEngine() {
        val alpha = 64
        underTest.alpha = alpha
        verify(engine).alpha = alpha
    }

    @Test
    fun returnsAlphaFromEngine() {
        val alpha = 128
        whenever(engine.alpha).thenReturn(alpha)
        assertEquals(alpha, underTest.alpha)
    }

    @Test
    fun setsColorFilterToRenderer() {
        val colorFilter: ColorFilter = mock()
        underTest.colorFilter = colorFilter
        verify(canvasRenderer).setColorFilter(colorFilter)
    }

    @Test
    fun opacityIsTranslucent() {
        assertEquals(PixelFormat.TRANSLUCENT, underTest.opacity)
    }

    @Test
    fun startsEngineOnStart() {
        underTest.start()
        verify(engine).start()
    }

    @Test
    fun stopsEngineOnStop() {
        underTest.stop()
        verify(engine).stop()
    }

    @Test
    fun returnsRunningStateFromEngine() {
        whenever(engine.isRunning).thenReturn(true)
        assertTrue(underTest.isRunning)
    }

    @Test
    fun forwardsNextFrameToEngine() {
        underTest.nextFrame()
        verify(engine).nextFrame()
    }

    @Test
    fun forwardsMakeFreshFrameToEngine() {
        underTest.makeFreshFrame()
        verify(engine).makeFreshFrame()
    }

    @Test
    fun forwardsMakeFreshFrameWithParticlesOffScreenToEngine() {
        underTest.makeFreshFrameWithParticlesOffscreen()
        verify(engine).makeFreshFrameWithParticlesOffscreen()
    }

    @Test
    fun forwardsSetDensityToScene() {
        val value = 1
        underTest.density = value
        verify(scene).density = value
    }

    @Test
    fun returnsDensityFromScene() {
        val value = 2
        whenever(scene.density).thenReturn(value)
        assertEquals(value, underTest.density)
    }

    @Test
    fun forwardsSetFrameDelayToScene() {
        val value = 3
        underTest.frameDelay = value
        verify(scene).frameDelay = value
    }

    @Test
    fun returnsFrameDelayFromScene() {
        val value = 4
        whenever(scene.frameDelay).thenReturn(value)
        assertEquals(value, underTest.frameDelay)
    }

    @Test
    fun forwardsSetParticleColorToScene() {
        val value = 5
        underTest.particleColor = value
        verify(scene).particleColor = value
    }

    @Test
    fun returnsParticleColorFromScene() {
        val value = 6
        whenever(scene.particleColor).thenReturn(value)
        assertEquals(value, underTest.particleColor)
    }

    @Test
    fun forwardsSetParticleRadiusRangeToScene() {
        val minRadius = 7f
        val maxRadius = 8f

        underTest.setParticleRadiusRange(minRadius, maxRadius)

        verify(scene).setParticleRadiusRange(minRadius, maxRadius)
    }

    @Test
    fun returnsParticleRadiusMaxFromScene() {
        val value = 9f
        whenever(scene.particleRadiusMax).thenReturn(value)
        assertEquals(value, underTest.particleRadiusMax)
    }

    @Test
    fun returnsParticleRadiusMinFromScene() {
        val value = 10f
        whenever(scene.particleRadiusMin).thenReturn(value)
        assertEquals(value, underTest.particleRadiusMin)
    }

    @Test
    fun forwardsSetLineColorToScene() {
        val value = 11
        underTest.lineColor = value
        verify(scene).lineColor = value
    }

    @Test
    fun returnsLineColorFromScene() {
        val value = 12
        whenever(scene.lineColor).thenReturn(value)
        assertEquals(value, underTest.lineColor)
    }

    @Test
    fun forwardsSetLineLengthToScene() {
        val value = 13f
        underTest.lineLength = value
        verify(scene).lineLength = value
    }

    @Test
    fun returnsLineLengthFromScene() {
        val value = 14f
        whenever(scene.lineLength).thenReturn(value)
        assertEquals(value, underTest.lineLength)
    }

    @Test
    fun forwardsSetLineThicknessToScene() {
        val value = 15f
        underTest.lineThickness = value
        verify(scene).lineThickness = value
    }

    @Test
    fun returnsLineThicknessFromScene() {
        val value = 16f
        whenever(scene.lineThickness).thenReturn(value)
        assertEquals(value, underTest.lineThickness)
    }

    @Test
    fun forwardsSetSpeedFactorToScene() {
        val value = 17f
        underTest.speedFactor = value
        verify(scene).speedFactor = value
    }

    @Test
    fun returnsSpeedFactorFromScene() {
        val value = 18f
        whenever(scene.speedFactor).thenReturn(value)
        assertEquals(value, underTest.speedFactor)
    }
}
